### ISSUE TITLE
implement new TreeScheduler

### ISSUE THREAD

**Issue by docmerlin** [influxdata/influxdb#15230]
## Requirements

Implement a scheduler that has the following methods:

```go
Stop()
Release(taskID ID)
Schedule(sch Schedulable) error
```

1) The scheduler should accept on creation an:
type ExecutorFunc func(ctx context.Context, id ID, scheduledAt time.Time) error and a type 

```go
Checkpointer interface {
	// Checkpoint saves the last checkpoint a ID has reached.
	Checkpoint(ctx context.Context, id ID, t time.Time) error
}
```

2) The scheduler should, after creation, automatically call ExecutorFunc, when a task should run as defined by its Schedulable.

3) the scheduler's should not be able to get into a state where blocks Release and Schedule indefinitely.

4) Schedule should add a Schedulable to being scheduled, and Release should remove a task from being scheduled.

5) Calling of ExecutorFunc should be serial in time on a per taskID basis.

## Design

The core of the scheduler is a btree keyed by time, a nonce, and a task ID, and a map keyed by task ID and containing a nonce and a time (called a uniqueness index from now on).  The map is to ensure task uniqueness in the tree, so we can replace or delete tasks in the tree.

Scheduling in the tree consists of a main loop that feeds a fixed set of workers, each with their own communication channel.  Distribution is handled by hashing the TaskID (to ensure uniform distribution) and then distributing over those channels evenly based on the hashed ID.  This is to ensure that all tasks of the same ID go to the same worker.

The workers call ExecutorFunc handle any errors and update the LastScheduled time internally and also via the Checkpointer.

### The main loop

The main loop waits on a time.Timer to grab the task with the minimum time.  Once it successfully grabs a task ready to trigger, it will start walking the btree from the item nearest 

### Putting a task into the scheduler

Adding a task to the scheduler acquires a write lock, grabs the task from the uniqueness map, and replaces the item in the uniqueness index and btree.  If new task would trigger sooner than the current soonest triggering task, it replaces the Timer when added to the scheduler.  Finally it releases the write lock.

### Removing a task from the scheduler

Removing a task from the scheduler acquires a write lock, deletes the task from the uniqueness index and from the btree, then releases the lock.  We do not have to readjust the time on delete, because, if the minimum task isn't ready yet, the main loop just resets the timer and keeps going.
